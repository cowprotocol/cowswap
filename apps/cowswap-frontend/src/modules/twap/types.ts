import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { TwapOrdersExecution } from './hooks/useTwapOrdersExecutions'

import { SafeTransactionParams } from '../../common/types'

// Read more: https://github.com/rndlabs/composable-cow#data-structure
export interface TWAPOrder {
  sellAmount: CurrencyAmount<Token>
  buyAmount: CurrencyAmount<Token>
  receiver: string
  numOfParts: number
  startTime: number
  timeInterval: number
  span: number
  appData: string
}

export interface TWAPOrderStruct {
  sellToken: string
  buyToken: string
  receiver: string
  partSellAmount: string
  minPartLimit: string
  // timeStart
  t0: number
  // numOfParts
  n: number
  // timeInterval
  t: number
  span: number
  appData: string
}

export enum TwapOrderStatus {
  WaitSigning = 'WaitSigning',
  Pending = 'Pending',
  Cancelling = 'Cancelling',
  Cancelled = 'Cancelled',
  Expired = 'Expired',
  Fulfilled = 'Fulfilled',
}

export interface TwapOrdersSafeData {
  conditionalOrderParams: ConditionalOrderParams
  safeTxParams: SafeTransactionParams
}

export interface TwapOrderExecutionInfo {
  executedSellAmount: string
  executedBuyAmount: string
  executedFeeAmount: string
}

export interface TwapOrderItem {
  order: TWAPOrderStruct
  status: TwapOrderStatus
  chainId: SupportedChainId
  executedDate?: string
  submissionDate: string
  safeAddress: string
  id: string
  safeTxParams?: SafeTransactionParams
  executionInfo: TwapOrdersExecution
}

export interface ConditionalOrderParams {
  staticInput: string
  salt: string
  handler: string
}

export interface TwapOrderInfo {
  id: string
  orderStruct: TWAPOrderStruct
  safeData: TwapOrdersSafeData
}

export type TwapOrdersAuthResult = { [key: string]: boolean | undefined }
