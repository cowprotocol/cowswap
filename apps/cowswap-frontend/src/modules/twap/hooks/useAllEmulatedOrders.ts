import { useMemo } from 'react'

import { UiOrderType } from '@cowprotocol/types'
import { useIsSafeApp, useWalletInfo } from '@cowprotocol/wallet'

import { Order } from 'legacy/state/orders/actions'
import { useOrders } from 'legacy/state/orders/hooks'

import { useEmulatedPartOrders } from './useEmulatedPartOrders'
import { useEmulatedTwapOrders } from './useEmulatedTwapOrders'
import { useTwapOrdersTokens } from './useTwapOrdersTokens'

export function useAllEmulatedOrders(): Order[] {
  const { chainId, account } = useWalletInfo()
  const twapOrdersTokens = useTwapOrdersTokens()
  const emulatedTwapOrders = useEmulatedTwapOrders(twapOrdersTokens)
  const emulatedPartOrders = useEmulatedPartOrders(twapOrdersTokens)
  const isSafeApp = useIsSafeApp()

  const twapOrders = useOrders(chainId, account, UiOrderType.TWAP)
  const discreteTwapOrders = useMemo(() => {
    return twapOrders.filter((order) => order.composableCowInfo?.isVirtualPart === false)
  }, [twapOrders])

  return useMemo(() => {
    if (!isSafeApp) return []

    return emulatedTwapOrders.concat(emulatedPartOrders).concat(discreteTwapOrders)
  }, [emulatedTwapOrders, emulatedPartOrders, discreteTwapOrders, isSafeApp])
}
