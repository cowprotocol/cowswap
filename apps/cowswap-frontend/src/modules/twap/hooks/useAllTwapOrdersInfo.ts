import { useEffect, useMemo, useRef, useState } from 'react'

import { TwapOrderInfo, TwapOrdersSafeData } from '../types'
import { getConditionalOrderId } from '../utils/getConditionalOrderId'
import { parseTwapOrderStruct } from '../utils/parseTwapOrderStruct'

export function useAllTwapOrdersInfo(ordersSafeData: TwapOrdersSafeData[]): TwapOrderInfo[] {
  const [allOrdersInfo, setAllOrdersInfo] = useState<Record<string, TwapOrderInfo>>({})
  const allOrdersInfoRef = useRef(allOrdersInfo)

  // eslint-disable-next-line react-hooks/refs
  allOrdersInfoRef.current = allOrdersInfo

  useEffect(() => {
    const parsed = parseOrdersSafeData(ordersSafeData)
    const ordersToAdd = parsed.filter((item) => !allOrdersInfoRef.current[item.id])

    if (!ordersToAdd.length) return

    const update = ordersToAdd.reduce<Record<string, TwapOrderInfo>>((acc, item) => {
      acc[item.id] = item

      return acc
    }, {})

    setAllOrdersInfo((state) => {
      return { ...state, ...update }
    })
  }, [ordersSafeData])

  return useMemo(() => Object.values(allOrdersInfo), [allOrdersInfo])
}

function parseOrdersSafeData(ordersSafeData: TwapOrdersSafeData[]): TwapOrderInfo[] {
  const ordersInfoMap = ordersSafeData.reduce<{ [id: string]: TwapOrderInfo }>((acc, data) => {
    try {
      const id = getConditionalOrderId(data.conditionalOrderParams)
      const existingOrder = acc[id]

      /**
       * There might be two Safe transactions with the same order inside.
       * But only one of them will be executed.
       *
       * For example, you propose a transaction with TWAP order and execute it.
       * Then, you propose another transaction with the same TWAP order.
       * After you realize that the proposed transaction is a duplicate, and you replace it or cancel.
       * In this case we should skip the second transaction, because the first one is already executed.
       */
      if (existingOrder?.safeData.safeTxParams.isExecuted) {
        return acc
      }

      const info = {
        id,
        orderStruct: parseTwapOrderStruct(data.conditionalOrderParams.staticInput),
        safeData: data,
      }

      acc[id] = info
    } catch {
      // Do nothing
    }

    return acc
  }, {})

  return Object.values(ordersInfoMap)
}
