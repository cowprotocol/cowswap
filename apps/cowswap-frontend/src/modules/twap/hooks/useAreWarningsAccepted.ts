import { useAtomValue } from 'jotai'

import { useIsNoImpactWarningAccepted } from 'modules/trade'

import { useIsFallbackHandlerRequired } from './useFallbackHandlerVerification'

import { twapOrdersSettingsAtom } from '../state/twapOrdersSettingsAtom'

export function useAreWarningsAccepted(): boolean {
  const { isFallbackHandlerSetupAccepted } = useAtomValue(twapOrdersSettingsAtom)
  const isNoImpactWarningAccepted = useIsNoImpactWarningAccepted()
  const isFallbackHandlerRequired = useIsFallbackHandlerRequired()

  if (isFallbackHandlerRequired) return isFallbackHandlerSetupAccepted

  return isNoImpactWarningAccepted
}
