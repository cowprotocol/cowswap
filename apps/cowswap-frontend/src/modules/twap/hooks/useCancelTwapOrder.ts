import { useAtomValue, useSetAtom } from 'jotai'
import { useCallback } from 'react'

import { useSafeAppsSdk } from '@cowprotocol/wallet'

import { Order } from 'legacy/state/orders/actions'

import { useComposableCowContract } from 'modules/advancedOrders/hooks/useComposableCowContract'

import type { OnChainCancellation } from 'common/hooks/useCancelOrder/onChainCancellation'
import { useGP2SettlementContract } from 'common/hooks/useContract'

import { cancelTwapOrderTxs, estimateCancelTwapOrderTxs } from '../services/cancelTwapOrderTxs'
import { setTwapOrderStatusAtom } from '../state/twapOrdersListAtom'
import { twapPartOrdersAtom } from '../state/twapPartOrdersAtom'
import { TwapOrderStatus } from '../types'

export function useCancelTwapOrder(): (twapOrderId: string, order: Order) => Promise<OnChainCancellation> {
  const twapPartOrders = useAtomValue(twapPartOrdersAtom)
  const setTwapOrderStatus = useSetAtom(setTwapOrderStatusAtom)
  const safeAppsSdk = useSafeAppsSdk()
  const settlementContract = useGP2SettlementContract()
  const composableCowContract = useComposableCowContract()

  return useCallback(
    async (twapOrderId: string, order: Order) => {
      if (!composableCowContract || !settlementContract || !safeAppsSdk) {
        throw new Error('Context is not full to cancel TWAP order')
      }

      const partOrder = twapPartOrders[twapOrderId]?.sort((a, b) => a.order.validTo - b.order.validTo)[0]
      const partOrderId = partOrder?.uid

      const context = { composableCowContract, settlementContract, orderId: twapOrderId, partOrderId }

      return {
        estimatedGas: await estimateCancelTwapOrderTxs(context),
        sendTransaction: (processCancelledOrder) => {
          return safeAppsSdk.txs.send({ txs: cancelTwapOrderTxs(context) }).then((res) => {
            const txHash = res.safeTxHash
            const sellTokenAddress = order.inputToken.address
            const sellTokenSymbol = order.inputToken.symbol

            setTwapOrderStatus(twapOrderId, TwapOrderStatus.Cancelling)
            processCancelledOrder({ txHash, orderId: twapOrderId, sellTokenAddress, sellTokenSymbol })
          })
        },
      }
    },
    [composableCowContract, settlementContract, safeAppsSdk, twapPartOrders, setTwapOrderStatus]
  )
}
