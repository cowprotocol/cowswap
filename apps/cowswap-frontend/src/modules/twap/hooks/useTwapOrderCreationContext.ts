import { ComposableCoW, Erc20 } from '@cowprotocol/abis'
import { useWalletInfo } from '@cowprotocol/wallet'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { Nullish } from 'types'

import { CURRENT_BLOCK_FACTORY_ADDRESS } from 'modules/advancedOrders'
import { useComposableCowContract } from 'modules/advancedOrders/hooks/useComposableCowContract'
import { useNeedsZeroApproval } from 'modules/zeroApproval'

import { useTokenContract } from 'common/hooks/useContract'
import { useNeedsApproval } from 'common/hooks/useNeedsApproval'
import { useTradeSpenderAddress } from 'common/hooks/useTradeSpenderAddress'

export interface TwapOrderCreationContext {
  composableCowContract: ComposableCoW
  needsApproval: boolean
  needsZeroApproval: boolean
  spender: string
  currentBlockFactoryAddress: string
  erc20Contract: Erc20
}

export function useTwapOrderCreationContext(
  inputAmount: Nullish<CurrencyAmount<Token>>
): TwapOrderCreationContext | null {
  const { chainId } = useWalletInfo()
  const composableCowContract = useComposableCowContract()
  const needsApproval = useNeedsApproval(inputAmount)
  const erc20Contract = useTokenContract(inputAmount?.currency.address)
  const spender = useTradeSpenderAddress()
  const needsZeroApproval = useNeedsZeroApproval(erc20Contract, spender, inputAmount)
  const currentBlockFactoryAddress = chainId ? CURRENT_BLOCK_FACTORY_ADDRESS[chainId] : null

  if (!composableCowContract || !erc20Contract || !spender || !currentBlockFactoryAddress) return null

  return { composableCowContract, erc20Contract, needsApproval, needsZeroApproval, spender, currentBlockFactoryAddress }
}
