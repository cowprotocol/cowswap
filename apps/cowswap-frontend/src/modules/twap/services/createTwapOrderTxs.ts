import { MaxUint256 } from '@ethersproject/constants'
import { MetaTransactionData } from '@safe-global/safe-core-sdk-types'

import { TwapOrderCreationContext } from '../hooks/useTwapOrderCreationContext'
import { ConditionalOrderParams, TWAPOrder } from '../types'

export function createTwapOrderTxs(
  order: TWAPOrder,
  paramsStruct: ConditionalOrderParams,
  context: TwapOrderCreationContext
): MetaTransactionData[] {
  const {
    composableCowContract,
    needsApproval,
    needsZeroApproval,
    erc20Contract,
    spender,
    currentBlockFactoryAddress,
  } = context

  const { sellAmount } = order

  const sellTokenAddress = sellAmount.currency.address
  const sellAmountAtoms = MaxUint256.toString()

  const txs: MetaTransactionData[] = []

  const createOrderTx = {
    to: composableCowContract.address,
    data: composableCowContract.interface.encodeFunctionData('createWithContext', [
      paramsStruct,
      currentBlockFactoryAddress,
      '0x',
      true,
    ]),
    value: '0',
    operation: 0,
  }

  txs.push(createOrderTx)

  if (!needsApproval) return txs

  const approveTx = {
    to: sellTokenAddress,
    data: erc20Contract.interface.encodeFunctionData('approve', [spender, sellAmountAtoms]),
    value: '0',
    operation: 0,
  }

  txs.unshift(approveTx)

  if (!needsZeroApproval) return txs

  const zeroApproveTx = {
    to: sellAmount.currency.address,
    data: erc20Contract.interface.encodeFunctionData('approve', [spender, '0']),
    value: '0',
    operation: 0,
  }

  txs.unshift(zeroApproveTx)

  return txs
}
