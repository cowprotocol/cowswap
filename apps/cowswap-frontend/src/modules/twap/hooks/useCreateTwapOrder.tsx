import { useAtomValue, useSetAtom } from 'jotai'
import { useCallback } from 'react'

import { orderAnalytics, twapConversionAnalytics } from '@cowprotocol/analytics'
import { OrderKind } from '@cowprotocol/cow-sdk'
import { UiOrderType } from '@cowprotocol/types'
import { useSafeAppsSdk, useWalletInfo } from '@cowprotocol/wallet'
import { CurrencyAmount, Token } from '@uniswap/sdk-core'

import { Nullish } from 'types'

import { updateAdvancedOrdersAtom, useAdvancedOrdersDerivedState } from 'modules/advancedOrders'
import { useAppData, useUploadAppData } from 'modules/appData'
import { emitPostedOrderEvent } from 'modules/orders'
import { getCowSoundSend } from 'modules/sounds'
import { useTradeConfirmActions, useTradePriceImpact } from 'modules/trade'
import { TradeFlowAnalyticsContext, tradeFlowAnalytics } from 'modules/trade/utils/analytics'

import { useConfirmPriceImpactWithoutFee } from 'common/hooks/useConfirmPriceImpactWithoutFee'

import { useExtensibleFallbackContext } from './useExtensibleFallbackContext'
import { useTwapOrderCreationContext } from './useTwapOrderCreationContext'

import { DEFAULT_TWAP_EXECUTION_INFO } from '../const'
import { createTwapOrderTxs } from '../services/createTwapOrderTxs'
import { extensibleFallbackSetupTxs } from '../services/extensibleFallbackSetupTxs'
import { twapOrderAtom } from '../state/twapOrderAtom'
import { addTwapOrderToListAtom } from '../state/twapOrdersListAtom'
import { TwapOrderItem, TwapOrderStatus } from '../types'
import { buildTwapOrderParamsStruct } from '../utils/buildTwapOrderParamsStruct'
import { getConditionalOrderId } from '../utils/getConditionalOrderId'
import { getErrorMessage } from '../utils/parseTwapError'
import { twapOrderToStruct } from '../utils/twapOrderToStruct'

export function useCreateTwapOrder() {
  const { chainId, account } = useWalletInfo()
  const twapOrder = useAtomValue(twapOrderAtom)
  const addTwapOrderToList = useSetAtom(addTwapOrderToListAtom)

  const { inputCurrencyAmount, outputCurrencyAmount } = useAdvancedOrdersDerivedState()

  const appDataInfo = useAppData()
  const safeAppsSdk = useSafeAppsSdk()
  const twapOrderCreationContext = useTwapOrderCreationContext(inputCurrencyAmount as Nullish<CurrencyAmount<Token>>)
  const extensibleFallbackContext = useExtensibleFallbackContext()

  const updateAdvancedOrdersState = useSetAtom(updateAdvancedOrdersAtom)

  const tradeConfirmActions = useTradeConfirmActions()
  const uploadAppData = useUploadAppData()

  const { priceImpact } = useTradePriceImpact()
  const { confirmPriceImpactWithoutFee } = useConfirmPriceImpactWithoutFee()

  return useCallback(
    async (fallbackHandlerIsNotSet: boolean) => {
      if (!chainId || !account) return
      if (
        !inputCurrencyAmount ||
        !outputCurrencyAmount ||
        !twapOrderCreationContext ||
        !extensibleFallbackContext ||
        !safeAppsSdk ||
        !appDataInfo ||
        !twapOrder
      )
        return

      const isPriceImpactConfirmed = await confirmPriceImpactWithoutFee(priceImpact)

      if (!isPriceImpactConfirmed) {
        return
      }

      const pendingTrade = {
        inputAmount: inputCurrencyAmount,
        outputAmount: outputCurrencyAmount,
      }

      const orderType = UiOrderType.TWAP
      const twapFlowAnalyticsContext: TradeFlowAnalyticsContext = {
        account,
        recipient: twapOrder.receiver,
        recipientAddress: twapOrder.receiver,
        marketLabel: [inputCurrencyAmount.currency.symbol, outputCurrencyAmount.currency.symbol].join(','),
        orderType,
      }

      try {
        const paramsStruct = buildTwapOrderParamsStruct(chainId, twapOrder)
        const orderId = getConditionalOrderId(paramsStruct)

        tradeConfirmActions.onSign(pendingTrade)
        tradeFlowAnalytics.placeAdvancedOrder(twapFlowAnalyticsContext)
        twapConversionAnalytics('posted', fallbackHandlerIsNotSet)

        const fallbackSetupTxs = fallbackHandlerIsNotSet
          ? await extensibleFallbackSetupTxs(extensibleFallbackContext)
          : []
        const createOrderTxs = createTwapOrderTxs(twapOrder, paramsStruct, twapOrderCreationContext)
        const { safeTxHash } = await safeAppsSdk.txs.send({ txs: [...fallbackSetupTxs, ...createOrderTxs] })

        const orderItem: TwapOrderItem = {
          order: twapOrderToStruct(twapOrder),
          status: TwapOrderStatus.WaitSigning,
          chainId,
          safeAddress: account,
          submissionDate: new Date().toISOString(),
          id: orderId,
          executionInfo: { confirmedPartsCount: 0, info: DEFAULT_TWAP_EXECUTION_INFO },
        }

        addTwapOrderToList(orderItem)

        getCowSoundSend().play()

        emitPostedOrderEvent({
          chainId,
          id: orderId,
          orderCreationHash: safeTxHash,
          kind: OrderKind.SELL,
          receiver: twapOrder.receiver,
          inputAmount: twapOrder.sellAmount,
          outputAmount: twapOrder.buyAmount,
          owner: account,
          uiOrderType: orderType,
        })

        orderAnalytics('Posted', orderType, 'Presign')

        uploadAppData({ chainId, orderId, appData: appDataInfo })
        updateAdvancedOrdersState({ recipient: null, recipientAddress: null })
        tradeConfirmActions.onSuccess(safeTxHash)
        tradeFlowAnalytics.sign(twapFlowAnalyticsContext)
        twapConversionAnalytics('signed', fallbackHandlerIsNotSet)
      } catch (error) {
        console.error('[useCreateTwapOrder] error', error)
        const errorMessage = getErrorMessage(error)
        tradeConfirmActions.onError(errorMessage)
        tradeFlowAnalytics.error(error, errorMessage, twapFlowAnalyticsContext)
        twapConversionAnalytics('rejected', fallbackHandlerIsNotSet)
      }
    },
    [
      chainId,
      account,
      inputCurrencyAmount,
      outputCurrencyAmount,
      twapOrderCreationContext,
      extensibleFallbackContext,
      safeAppsSdk,
      appDataInfo,
      twapOrder,
      confirmPriceImpactWithoutFee,
      priceImpact,
      tradeConfirmActions,
      addTwapOrderToList,
      uploadAppData,
      updateAdvancedOrdersState,
    ]
  )
}
