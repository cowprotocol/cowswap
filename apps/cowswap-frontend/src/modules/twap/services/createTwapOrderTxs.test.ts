import { COW, WETH_SEPOLIA } from '@cowprotocol/common-const'
import { SupportedChainId } from '@cowprotocol/cow-sdk'
import { CurrencyAmount } from '@uniswap/sdk-core'

import { COMPOSABLE_COW_ADDRESS, CURRENT_BLOCK_FACTORY_ADDRESS } from 'modules/advancedOrders'
import { getAppData } from 'modules/appData'

import { createTwapOrderTxs } from './createTwapOrderTxs'

import { TwapOrderCreationContext } from '../hooks/useTwapOrderCreationContext'
import { TWAPOrder } from '../types'
import { buildTwapOrderParamsStruct } from '../utils/buildTwapOrderParamsStruct'

jest.mock('modules/permit')

const APP_DATA_HASH = getAppData().appDataKeccak256

const chainId = SupportedChainId.SEPOLIA

const order: TWAPOrder = {
  sellAmount: CurrencyAmount.fromRawAmount(COW[SupportedChainId.SEPOLIA], 100_000_000_000),
  buyAmount: CurrencyAmount.fromRawAmount(WETH_SEPOLIA, 200_000),
  receiver: '0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6',
  numOfParts: 3,
  startTime: 1684764716,
  timeInterval: 600,
  span: 0,
  appData: APP_DATA_HASH,
}

const CREATE_COW_TX_DATA = '0xCREATE_COW_TX_DATA'
const APPROVE_TX_DATA = '0xAPPROVE_TX_DATA'

describe('Create TWAP order', () => {
  let context: TwapOrderCreationContext
  let createCowFn: jest.Mock
  let approveFn: jest.Mock

  beforeEach(() => {
    jest.spyOn(Date, 'now').mockImplementation(() => 1497076708000)

    createCowFn = jest.fn().mockReturnValue(CREATE_COW_TX_DATA)
    approveFn = jest.fn().mockReturnValue(APPROVE_TX_DATA)

    context = {
      composableCowContract: {
        interface: { encodeFunctionData: createCowFn },
        address: COMPOSABLE_COW_ADDRESS[chainId],
      } as any,
      needsApproval: false,
      needsZeroApproval: false,
      spender: '0xB4FBF271143F4FBf7B91A5ded31805e42b222222',
      currentBlockFactoryAddress: CURRENT_BLOCK_FACTORY_ADDRESS[chainId],
      erc20Contract: { interface: { encodeFunctionData: approveFn } } as any,
    }
  })

  it('When sell token is approved, then should generate only creation transaction', () => {
    const paramsStruct = buildTwapOrderParamsStruct(chainId, order)
    const result = createTwapOrderTxs(order, paramsStruct, { ...context, needsApproval: false })

    expect(createCowFn).toHaveBeenCalledTimes(1)
    expect(createCowFn.mock.calls[0]).toMatchSnapshot()

    expect(result.length).toBe(1)
    expect(result).toMatchSnapshot()
  })

  it('When sell token is NOT approved, then should generate approval and creation transactions', () => {
    const paramsStruct = buildTwapOrderParamsStruct(chainId, order)
    const result = createTwapOrderTxs(order, paramsStruct, { ...context, needsApproval: true })

    expect(createCowFn).toHaveBeenCalledTimes(1)
    expect(createCowFn.mock.calls[0]).toMatchSnapshot()

    expect(approveFn).toHaveBeenCalledTimes(1)
    expect(approveFn.mock.calls[0]).toMatchSnapshot()

    expect(result.length).toBe(2)
    expect(result).toMatchSnapshot()
  })

  it('When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions', () => {
    const paramsStruct = buildTwapOrderParamsStruct(chainId, order)
    const result = createTwapOrderTxs(order, paramsStruct, { ...context, needsApproval: true, needsZeroApproval: true })

    expect(createCowFn).toHaveBeenCalledTimes(1)
    expect(createCowFn.mock.calls[0]).toMatchSnapshot()

    expect(approveFn).toHaveBeenCalledTimes(2)
    expect(approveFn.mock.calls[0]).toMatchSnapshot()
    expect(approveFn.mock.calls[1]).toMatchSnapshot()

    expect(result.length).toBe(3)
    expect(result).toMatchSnapshot()
  })
})
