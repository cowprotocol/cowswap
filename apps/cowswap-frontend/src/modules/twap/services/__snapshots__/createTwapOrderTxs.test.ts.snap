// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 1`] = `
[
  "createWithContext",
  [
    {
      "handler": "0x6cF1e9cA41f7611dEf408122793c358a3d11E5a5",
      "salt": "0x00000000000000000000000000000000000000000000000000000015c90b9b2a",
      "staticInput": "0x0000000000000000000000000625afb445c3b6b7b929342a04a22599fd5dbb59000000000000000000000000fff9976782d46cc05630d1f6ebab18b2324d6b14000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000000000000000000000000000000000007c2d24d55000000000000000000000000000000000000000000000000000000000001046a00000000000000000000000000000000000000000000000000000000646b782c0000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000025800000000000000000000000000000000000000000000000000000000000000003f13237929334d44bdb32b189d1644965915936e62004fb77ec7bc861cddc7a4",
    },
    "0x52eD56Da04309Aca4c3FECC595298d80C2f16BAc",
    "0x",
    true,
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 2`] = `
[
  "approve",
  [
    "0xB4FBF271143F4FBf7B91A5ded31805e42b222222",
    "115792089237316195423570985008687907853269984665640564039457584007913129639935",
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 3`] = `
[
  "approve",
  [
    "0xB4FBF271143F4FBf7B91A5ded31805e42b222222",
    "0",
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved AND token needs zero approval, then should generate 2 approvals and creation transactions 4`] = `
[
  {
    "data": "0xAPPROVE_TX_DATA",
    "operation": 0,
    "to": "0x0625aFB445C3B6B7B929342a04A22599fd5dBB59",
    "value": "0",
  },
  {
    "data": "0xAPPROVE_TX_DATA",
    "operation": 0,
    "to": "0x0625aFB445C3B6B7B929342a04A22599fd5dBB59",
    "value": "0",
  },
  {
    "data": "0xCREATE_COW_TX_DATA",
    "operation": 0,
    "to": "0xfdaFc9d1902f4e0b84f65F49f244b32b31013b74",
    "value": "0",
  },
]
`;

exports[`Create TWAP order When sell token is NOT approved, then should generate approval and creation transactions 1`] = `
[
  "createWithContext",
  [
    {
      "handler": "0x6cF1e9cA41f7611dEf408122793c358a3d11E5a5",
      "salt": "0x00000000000000000000000000000000000000000000000000000015c90b9b2a",
      "staticInput": "0x0000000000000000000000000625afb445c3b6b7b929342a04a22599fd5dbb59000000000000000000000000fff9976782d46cc05630d1f6ebab18b2324d6b14000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000000000000000000000000000000000007c2d24d55000000000000000000000000000000000000000000000000000000000001046a00000000000000000000000000000000000000000000000000000000646b782c0000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000025800000000000000000000000000000000000000000000000000000000000000003f13237929334d44bdb32b189d1644965915936e62004fb77ec7bc861cddc7a4",
    },
    "0x52eD56Da04309Aca4c3FECC595298d80C2f16BAc",
    "0x",
    true,
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved, then should generate approval and creation transactions 2`] = `
[
  "approve",
  [
    "0xB4FBF271143F4FBf7B91A5ded31805e42b222222",
    "115792089237316195423570985008687907853269984665640564039457584007913129639935",
  ],
]
`;

exports[`Create TWAP order When sell token is NOT approved, then should generate approval and creation transactions 3`] = `
[
  {
    "data": "0xAPPROVE_TX_DATA",
    "operation": 0,
    "to": "0x0625aFB445C3B6B7B929342a04A22599fd5dBB59",
    "value": "0",
  },
  {
    "data": "0xCREATE_COW_TX_DATA",
    "operation": 0,
    "to": "0xfdaFc9d1902f4e0b84f65F49f244b32b31013b74",
    "value": "0",
  },
]
`;

exports[`Create TWAP order When sell token is approved, then should generate only creation transaction 1`] = `
[
  "createWithContext",
  [
    {
      "handler": "0x6cF1e9cA41f7611dEf408122793c358a3d11E5a5",
      "salt": "0x00000000000000000000000000000000000000000000000000000015c90b9b2a",
      "staticInput": "0x0000000000000000000000000625afb445c3b6b7b929342a04a22599fd5dbb59000000000000000000000000fff9976782d46cc05630d1f6ebab18b2324d6b14000000000000000000000000b4fbf271143f4fbf7b91a5ded31805e42b2208d600000000000000000000000000000000000000000000000000000007c2d24d55000000000000000000000000000000000000000000000000000000000001046a00000000000000000000000000000000000000000000000000000000646b782c0000000000000000000000000000000000000000000000000000000000000003000000000000000000000000000000000000000000000000000000000000025800000000000000000000000000000000000000000000000000000000000000003f13237929334d44bdb32b189d1644965915936e62004fb77ec7bc861cddc7a4",
    },
    "0x52eD56Da04309Aca4c3FECC595298d80C2f16BAc",
    "0x",
    true,
  ],
]
`;

exports[`Create TWAP order When sell token is approved, then should generate only creation transaction 2`] = `
[
  {
    "data": "0xCREATE_COW_TX_DATA",
    "operation": 0,
    "to": "0xfdaFc9d1902f4e0b84f65F49f244b32b31013b74",
    "value": "0",
  },
]
`;
